src/cocaine-app/sched/__init__.py

```diff
@@ -0,0 +1,417 @@
+
+        if not self.params.get(starter_name, {}).get('enabled', False):
+            logger.info("Starter {} is disabled".format(starter_name))
+            return
+
+        timer = periodic_timer(seconds=period_val)
+
+        def _starter_periodic_func():
+            try:
+                logger.info('Starting {}'.format(starter_name))
+
+                with sync_manager.lock(lock_name, blocking=False):
+                    starter_func()
+
+            except LockFailedError:
+                logger.info('Another {} is already running under {} lock'.format(starter_func.__name__, lock_name))
+            except:
+                logger.exception('Failed to {}'.format(starter_name))
+            finally:
+                logger.info('{} finished'.format(starter_name))
+                self.__tq.add_task_at(
+                    starter_name,
+                    timer.next(),
+                    _starter_periodic_func)
+
+        self.__tq.add_task_at(
+            starter_name,
+            timer.next(),
+            _starter_periodic_func
+        )
+
+        starter = {
+            "starter_func": starter_func,
+            "starter_name": starter_name,
+            "starter_lock": lock_name,
+            "starter_timer": timer,
+        }
+
+        self.starters.append(starter)
+
+        logger.info("Registered starter function")
+
+    def update_resource_stat(self):
+
+        # update jobs list
+        active_jobs = self.job_processor.job_finder.jobs(
+            statuses=jobs.Job.ACTIVE_STATUSES,
+            sort=False,
+        )
+
+        # update job count per each type
+        self.job_count = Counter()
+
+        # update host stat
+        self.res = defaultdict(list)
+
+        errors = []
+
+        for job in active_jobs:
+            self.job_count[job.type] += 1
+            res_demand = self.convert_resource_representation(job.resources, job._involved_groups, job.type, errors)
+            self.add_to_resource_stat(res_demand, job.id)
+
+        if len(errors) != 0:
+            logger.warn("Errors during resource aggregation are {}".format(set(errors)))
+
+    def convert_resource_representation(self, resources, groups, job_type, errors=None):
+        """
+        The function takes old resources (equivalent to job.resources), "resource_limits" for each job type from config
+        and produces a new resource representation
+        :param resources:  Dict of res_type where each value is a list of node addr or tuples (node_addr, fs)
+        :param groups: List of groups ids
+        :param job_type: job_type
+        :param errors: list of strings
+        :return: a dict where keys are tuples (res_type, group) or (res_type, node_addr) or (res_type, node_addr, fs_id)
+                and values are consumption level
+        """
+        res_demand = {}
+
+        for g in groups:
+            res_demand[("Group", g)] = 100
```

Comment:
Fixed in https://github.com/yandex/mastermind/pull/417/commits/e7a0f53468278b6dc781abea9c82cb3cb7350f4e